utils/transfer/src/progress.rs

```diff
@@ -0,0 +1,93 @@
+use crate::error::Error;
+use crate::{abortable_sink, abortable_stream, TransferSink, TransferStream};
+use crate::{TransferContext, TransferData};
+use futures::{SinkExt, StreamExt, TryFutureExt};
+use tokio::task::spawn_local;
+
+type Stream = TransferStream<TransferData, Error>;
+
+/// Wraps a stream to report progress.
+/// The `report` function is called with the current offset and the total size.
+/// The total size is 0 if the size is unknown. (For example, when the source is a directory.)
+/// The reporting function should not block as it will block the transfer.
+pub fn wrap_stream_with_progress_reporting<F>(
```

Comment:
I would make this member function on `TransferStream`

Reply:
The problem TransferStream is generic (although I think at all places it is `TransferStream<TransferData,  ya_transfer::error>` but for some reason this needs to be an open type) so to accomodate that I would need to either do:

```rust
impl TransferStream<TransferData, Error> {
```
or
```rust
impl <E> TransfreStream<TransferData, E> {
```
to accommodate a generic error type _or_ to better support general usage, I would need to create a trait to get the data length like:
```rust
trait Measurable {
  get_len(&self) -> u64
}
impl <T: Measurable, R> TransferStream<T, E> {
...
}
impl Measurable for Transfer Data {
...
```
I am not sure which version would be better or if it is worth it.

Reply:
It seems not worth if it requires so much work.